proposal/nested-map-functions.md

```diff
@@ -0,0 +1,389 @@
+> function returns `false` rather than throwing an error.
+
+* If `$map` is not a map, throw an error.
+
+* Let `child` be `$map`.
+
+* Let `keys` be a list containing `$key` followed by the elements of `$keys`.
+
+* For each element `key` in `keys`:
+
+  * If `child` is not a map, return `false`.
+
+  * If `child` contains a key that's `==` to `key`, set `child` to the value
+    associated with that key. Otherwise, return `false`.
+
+* Return `true`.
+
+### `set()`
+
+> Note: For consistency with other functions whose multi-key overloads were
+> added after their single-key versions, `set()` is defined to have a separate
+> single-key overload and multi-key overload.
+
+* ```
+  set($map, $key, $value)
+  ```
+
+  > Intuitively, `set($map, $key, $value)` is equivalent to `merge($map, ($key: $value))`.
+
+  * If `$map` is not a map, throw an error.
+
+  * Let `map` be a copy of `$map`.
+
+  * If `map` has a key that's `==` to `$key`, remove it and its associated value.
+
+  * Associate `$key` with `$value` in `map`.
+
+  * Return `map`.
+
+* ```
+  set($map, $args...)
+  ```
+
+  > Intuitively, `set($map, $key1, $key2, $value)` is equivalent to `set($map,
+  > $key1, set(get($map, $key1), $key2, $value))` with the exception that if any
+  > intermediate value isn't set or isn't a map it's replaced with a map.
+
+  * If `$map` is not a map, throw an error.
+
+  * If `$args` has fewer than three elements, throw an error.
+
+  * Let `map` be a copy of `$map`.
+
+  * Let `key` be the first element of `$args`.
+
+  * Let `remaining` be the slice of all elements in `$args` except the first.
+
+  * If `map` has a key that's `==` to `key`:
+
+    * Remove that key and its associated value from `map`.
+
+    * Let `child` be the value that was associated with that key if that value
+      is a map, or an empty map otherwise.
+
+  * Otherwise:
+
+    * Let `child` be an empty map.
+
+  * Let `new-child` be the result of calling `set()` with `child` as the first
+    argument and the elements of `remaining` as the remaining arguments.
+
+  * Associate `key` with `new-child` in `map`.
+
+  * Return `map`.
+
+### `merge()`
+
+This proposal adds a new overload to the existing `merge()` function with lower
+priority than the existing signature.
+
+> This means that the new overload is only called if the existing signature
+> doesn't match.
+
+This proposal adds a new overload to the existing `merge()` function:
+
+```
+merge($map1, $args...)
+```
+
+> Intuitively, `map.merge($map1, $keys..., $map2)` is equivalent to
+> `map.set($map1, $keys..., map.merge(map.get($map1, $keys...), $map2))`.
+
+* If `$args` is empty, return `$map1`.
+
+* Let `map2` be the last element of `$args`.
+
+* If either `$map1` or `map2` is not a map, throw an error.
+
+* Let `keys` be a slice of all elements in `$args` except the last.
+
+* Let `sub` be the result of calling `get()` with `$map1` as the first
+  argument and the contents of `keys` as the remaining arguments.
```

Comment:
What happens if `keys` is empty?

Reply:
That's not possible here, because that would end up calling the first overload. I'll add an explicit error for it, though.